src/Core/Animation/KeyFramedValue.hpp

```diff
@@ -0,0 +1,215 @@
+ * of KeyFramedValue::KeyFrame and returns the value for a given time t
+ * (see for instance Ra::Core::Animation::linearInterpolate).
+ */
+template <typename VALUE_TYPE>
+class KeyFramedValue : public KeyFramedValueBase
+{
+  public:
+    /// The type for a keyframe.
+    using KeyFrame = std::pair<Scalar, VALUE_TYPE>;
+
+    /// The type for the keyframes container.
+    using KeyFrames = std::vector<KeyFrame>;
+
+    /// The type for interpolators.
+    using Interpolator = std::function<VALUE_TYPE( const KeyFrames& /*frames*/, Scalar /*t*/ )>;
+
+    /**
+     * Creates a KeyFramedValue from a first keyframe.
+     * \param frame the first keyframe value.
+     * \param t the first keyframe point in time.
+     * \param interpolator the function used to interpolate between keyframes.
+     */
+    KeyFramedValue( VALUE_TYPE frame,
+                    Scalar t,
+                    Interpolator interpolator = []( const KeyFrames& frames,
+                                                    Scalar ) { return frames.begin()->second; } ) :
+        KeyFramedValueBase(),
+        m_interpolator( interpolator ) {
+        insertKeyFrame( t, frame );
+    }
+
+    KeyFramedValue( const KeyFramedValue& keyframe ) = default;
+
+    inline KeyFramedValue& operator=( const KeyFramedValue& keyframe ) = default;
+
+    inline size_t size() const override final { return m_keyframes.size(); }
+
+    inline std::vector<Scalar> getTimes() const override final {
+        std::vector<Scalar> time( m_keyframes.size() );
+#pragma omp parallel for
+        for ( int i = 0; i < m_keyframes.size(); ++i )
+        {
+            time[i] = m_keyframes[i].first;
+        }
+        return time;
+    }
+
+    /// \name KeyFrame Management
+    /// \{
+
+    /**
+     * \returns the \p i-th keyframe.
+     */
+    inline const KeyFrame& getKeyFrame( size_t i ) { return m_keyframes[i]; }
+
+    /**
+    ￼ * Inserts a new keyframe with value \p frame at time \p t.
+    ￼ * \note If a keyframe already exists for \p t, it will be overwritten.
+    ￼ */
+    inline void insertKeyFrame( const Scalar& t, const VALUE_TYPE& frame ) {
+        KeyFrame kf( t, frame );
+        auto upper = std::upper_bound(
+            m_keyframes.begin(), m_keyframes.end(), kf, []( const auto& a, const auto& b ) {
+                return a.first < b.first;
+            } );
+        auto lower = upper;
+        --lower;
+        if ( upper == m_keyframes.end() ) { m_keyframes.insert( upper, kf ); }
+        else if ( lower == m_keyframes.end() )
+        { m_keyframes.insert( m_keyframes.begin(), kf ); }
+        else if ( Math::areApproxEqual( lower->first, t ) )
+        { lower->second = frame; }
+        else
+        { m_keyframes.insert( upper, kf ); }
+    }
+
+    /**
+    ￼ * Inserts a keyframe at time \p t corresponding to the value interpolated at \p t.
+    ￼ */
+    inline void insertKeyFrame( const Scalar& t ) override final { insertKeyFrame( t, at( t ) ); }
+
```

Comment:
See comment above (rename refineKeyFrame ?) and, as for `at`, this method could be templated by the interpolator. One can refine a key-frame set using an interpolator different from the one used to sample values at time t. (eg. spline when refining, linear when sampling)